### PR TITLE
Issue/21799

### LINKED ISSUE #21799: 🐛 Bug Report: Issues with Sidebar Navigation to Catalog
### 📜 Description

1. SidebarSubmenuItem navigation is overriding filters[user]=owned when navigating in the catalog
2. After clicking link in Sidebar to catalog, if you change the filters on the page the link still appears active and clicking the link in the Sidebar again updates the URL but doesn't re-render the page

### 👍 Expected behavior

Clicking on the item in the Sidebar should take you to the link as it is given.

### 👎 Actual Behavior with Screenshots

**First Issue**
When clicking on the link in the Sidebar to the catalog where the user owns something, it works as expected and takes you to the owned view.

However, if you first click on a link in the Sidebar on an item where a user doesn't own something, for example (the owned filter is there but currently none of our users can own a team): 
`<SidebarSubmenuItem
              title="Owning Teams"
              to="/catalog?filters[kind]=group&filters[user]=owned"
              exact
            />`
And then you click on a link in the Sidebar where the user should own something
` <SidebarSubmenuItem
              title="Components"
              to="catalog?filters[kind]=component&filters[user]=owned"
              exact
            />
`
The URL switches to all anyway

**Second Issue**
Page does not update back to original link

Video shows both issues:

https://github.com/backstage/backstage/assets/57731239/163a972e-9d0f-49e3-bf59-64b3553483f7

### 👟 Reproduction steps

In Root.tsx

>         <SidebarItem icon={CategoryIcon} text="Catalog">
>          {' '}
>          <SidebarSubmenu title="Catalog">
>            <SidebarSubmenuItem
>              title="Components"
>              to="catalog?filters[kind]=component&filters[user]=owned"
>              exact
>            />
>            <SidebarSubmenuItem
>              title="APIs"
>              to="catalog?filters[kind]=api&filters[user]=owned"
>              exact
>            />
>            <SidebarSubmenuItem
>              title="Owning Teams"
>              to="catalog?filters[kind]=group&filters[user]=owned"
>              exact
>            />
>          </SidebarSubmenu>{' '}
>        </SidebarItem>

**First Issue**
Use Sidebar to go to link where user doesn't own anything, then use Sidebar to go to link where user does own something.

**Second Issue**
1. Use Sidebar to go to one of the links
2. Change the Kind or Type on the catalog page
3. Click the same link in the Sidebar as you did in step 1

### 📃 Provide the context for the Bug.

We want the links to always filter to the owned view if users own something in that category

### 🖥️ Your Environment

Bug exists on Chrome, Firefox, and Edge
Output of yarn backstage-cli info
@backstage/app-defaults                                          1.4.5
  @backstage/backend-app-api                                       0.4.5, 0.5.8
  @backstage/backend-common                                        0.18.5, 0.19.9
  @backstage/backend-dev-utils                                     0.1.2
  @backstage/backend-openapi-utils                                 0.1.0
  @backstage/backend-plugin-api                                    0.5.4, 0.6.7
  @backstage/backend-tasks                                         0.5.12
  @backstage/backend-test-utils                                    0.2.8
  @backstage/catalog-client                                        1.4.6
  @backstage/catalog-model                                         1.4.3
  @backstage/cli-common                                            0.1.13
  @backstage/cli-node                                              0.1.4, 0.2.0
  @backstage/cli                                                   0.22.13, 0.24.0
  @backstage/config-loader                                         1.5.3
  @backstage/config                                                1.1.1
  @backstage/core-app-api                                          1.11.1
  @backstage/core-components                                       0.12.4, 0.13.8
  @backstage/core-plugin-api                                       1.8.0
  @backstage/dev-utils                                             1.0.24
  @backstage/e2e-test-utils                                        0.1.0
  @backstage/errors                                                1.2.3
  @backstage/eslint-plugin                                         0.1.3
  @backstage/frontend-plugin-api                                   0.3.0
  @backstage/integration-aws-node                                  0.1.8
  @backstage/integration-react                                     1.1.21
  @backstage/integration                                           1.7.2
  @backstage/plugin-analytics-module-ga4                           0.1.6
  @backstage/plugin-api-docs                                       0.10.1
  @backstage/plugin-app-backend                                    0.3.55
  @backstage/plugin-app-node                                       0.1.7
  @backstage/plugin-auth-backend-module-gcp-iap-provider           0.2.1
  @backstage/plugin-auth-backend-module-github-provider            0.1.4
  @backstage/plugin-auth-backend-module-gitlab-provider            0.1.4
  @backstage/plugin-auth-backend-module-google-provider            0.1.4
  @backstage/plugin-auth-backend-module-oauth2-provider            0.1.4
  @backstage/plugin-auth-backend                                   0.20.0
  @backstage/plugin-auth-node                                      0.2.19, 0.4.1
  @backstage/plugin-catalog-backend-module-aws                     0.3.1
  @backstage/plugin-catalog-backend-module-scaffolder-entity-model 0.1.4
  @backstage/plugin-catalog-backend                                1.15.0
  @backstage/plugin-catalog-common                                 1.0.18
  @backstage/plugin-catalog-graph                                  0.3.1
  @backstage/plugin-catalog-import                                 0.10.3
  @backstage/plugin-catalog-node                                   1.5.0
  @backstage/plugin-catalog-react                                  1.9.1
  @backstage/plugin-catalog                                        1.15.1
  @backstage/plugin-events-node                                    0.2.16
  @backstage/plugin-github-actions                                 0.6.8
  @backstage/plugin-home-react                                     0.1.5
  @backstage/plugin-home                                           0.5.11
  @backstage/plugin-kubernetes-common                              0.7.1
  @backstage/plugin-org                                            0.6.17
  @backstage/plugin-permission-backend                             0.5.30
  @backstage/plugin-permission-common                              0.7.10
  @backstage/plugin-permission-node                                0.7.18
  @backstage/plugin-permission-react                               0.4.17
  @backstage/plugin-proxy-backend                                  0.4.5
  @backstage/plugin-scaffolder-backend                             1.19.1
  @backstage/plugin-scaffolder-common                              1.4.3
  @backstage/plugin-scaffolder-node                                0.2.8
  @backstage/plugin-scaffolder-react                               1.6.1
  @backstage/plugin-scaffolder                                     1.16.1
  @backstage/plugin-search-backend-module-catalog                  0.1.11
  @backstage/plugin-search-backend-module-pg                       0.5.16
  @backstage/plugin-search-backend-module-techdocs                 0.1.11
  @backstage/plugin-search-backend-node                            1.2.11
  @backstage/plugin-search-backend                                 1.4.7
  @backstage/plugin-search-common                                  1.2.8
  @backstage/plugin-search-react                                   1.7.3
  @backstage/plugin-search                                         1.4.3
  @backstage/plugin-tech-radar                                     0.6.10
  @backstage/plugin-techdocs-backend                               1.9.0
  @backstage/plugin-techdocs-module-addons-contrib                 1.1.2
  @backstage/plugin-techdocs-node                                  1.10.0
  @backstage/plugin-techdocs-react                                 1.1.13
  @backstage/plugin-techdocs                                       1.9.1
  @backstage/plugin-user-settings                                  0.7.13
  @backstage/release-manifests                                     0.0.10, 0.0.11
  @backstage/test-utils                                            1.4.5
  @backstage/theme                                                 0.2.19, 0.4.4
  @backstage/types                                                 1.1.1
  @backstage/version-bridge                                        1.0.7

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, but I'm happy to collaborate on a PR with someone else

### PR DESCRIPTION
**Pull request by mattprinold**
## Hey, I just made a Pull Request!

Fixes #21799 
- UserListPicker now only overrides the user filter query parameters if the kind filter is either group or user
- Adds optional prop **reloadDocument** that can be passed to SidebarSubmenuItem and SidebarItem that forces a remount of the page on click, resetting and state for that route. 

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
I'm not on the catalog maintainers team, but note that there are spell checker warnings on the changeset. Well, some about spelling as such, but some that also just needs to be added in ``` `backticks` ``` because they are technical identifiers

**Review by mikeyhc — APPROVED** [review]

**Comment by mattprinold** [comment]
@mikeyhc This keeps on failing on the accessibility tests and is blocking auto merge though looking at the failure it doesn't seem related to this work. Is there an issue in the master branch with this? I've rebased this branch so its inline with master.

**Comment by mattprinold** [comment]
@mikeyhc all checks passing now

**Comment by mikeyhc** [comment]
You have my approval, but someone from @backstage/maintainers will need to sign off

File: .changeset/popular-ladybugs-pay.md
Comment (freben) [inline .changeset/popular-ladybugs-pay.md:6]: this really should be two different changesets (could have been two PRs as well, but no worries)

File: plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx
Comment (freben) [inline plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx:186]: This removed the `filterCounts[selectedUserFilter] === 0` check, was that intentional?

Comment (mattprinold) [reply plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx:186]: @freben Yes that is intentional. As per the request the the selectedUserFilter should only automatically overide to `all` when the kind filter is `User` or `Group` for all other kinds the end user should be able to see if they dont have any entities that match `starred` or `owned`.

File: .changeset/popular-ladybugs-pay.md
Comment (mattprinold) [reply .changeset/popular-ladybugs-pay.md:6]: @freben Ill split this into two changesets. I did ask on the issue thread when i picked this up if it should be split into two separate PR's but i received no response.

Comment (mattprinold) [reply .changeset/popular-ladybugs-pay.md]: done

File: plugins/catalog-react/src/components/UserListPicker/UserListPicker.test.tsx
Comment (vinzscam) [inline plugins/catalog-react/src/components/UserListPicker/UserListPicker.test.tsx:483]: please don't use `MockEntityListContextProvider`. I know it will make testing easier but it will mock a lot of logic we don't want to mock here.

File: plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx
Comment (vinzscam) [reply plugins/catalog-react/src/components/UserListPicker/UserListPicker.tsx:186]: are we sure of this approach? Have we evaluated if it's possible to strip out the filter in `EntityListProvider` before api calling to catalog instead?

**Review by vinzscam — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
